File: typescript/examples/paper-to-images.ts
#!/usr/bin/env node

/**
 * Example: Converting Paper to Compressed Images
 * 
 * This example demonstrates how to compress text and render it as images
 * for vision model consumption.
 */

import { readFileSync, writeFileSync, existsSync } from 'fs';
import { join } from 'path';
import { StatisticalFilter, StatisticalFilterConfig, OutputFormat } from '../src';

async function main() {
  console.log('📄 Converting Paper to Compressed Images\n');

  // Read original paper
  const paperPath = join(__dirname, '../../benchmarks/datasets/arxiv_markdown/1505.07818.md');
  
  if (!existsSync(paperPath)) {
    console.error(`❌ Paper file not found: ${paperPath}`);
    process.exit(1);
  }
  
  const text = readFileSync(paperPath, 'utf-8');

  console.log('📝 Original paper:');
  console.log(`  File: ${paperPath}`);
  console.log(`  Size: ${text.length} chars`);
  console.log(`  Words: ${text.split(/\s+/).length}`);
  console.log();

  const sep = '='.repeat(80);

  // Configuration for different compression levels
  const configs: Array<[number, string, string]> = [
    [0.3, 'aggressive', '30% compression'],
    [0.5, 'balanced', '50% compression (default)'],
    [0.7, 'light', '70% compression'],
  ];

  for (const [ratio, label, description] of configs) {
    console.log(sep);
    console.log(`Processing: ${label.toUpperCase()} - ${description}`);
    console.log(sep);

    const config: StatisticalFilterConfig = {
      compressionRatio: ratio,
      idfWeight: 0.3,
      positionWeight: 0.2,
      posWeight: 0.2,
      entityWeight: 0.2,
      entropyWeight: 0.1,
      enableProtectionMasks: true,
      enableContextualStopwords: true,
      preserveNegations: true,
      preserveComparators: true,
      domainTerms: [],
      minGapBetweenCritical: 3,
    };

    const filter = new StatisticalFilter(config);
    const compressed = filter.compress(text);
    
    // Calculate token estimates
    const originalTokens = Math.floor(text.length / 4);
    const compressedTokens = Math.floor(compressed.length / 4);
    const compressionRatio = compressedTokens / originalTokens;
    const tokensRemoved = originalTokens - compressedTokens;

    console.log('✅ Compression complete:');
    console.log(`  Original tokens: ${originalTokens}`);
    console.log(`  Compressed tokens: ${compressedTokens}`);
    console.log(`  Compression ratio: ${(compressionRatio * 100).toFixed(1)}%`);
    console.log(`  Tokens saved: ${tokensRemoved}`);
    console.log(`  Token savings: ${((1 - compressionRatio) * 100).toFixed(1)}%`);

    // Save compressed text
    const textFilename = `rnn_paper_${label}_compressed.txt`;
    writeFileSync(textFilename, compressed);
    console.log(`  📝 Text saved: ${textFilename}`);

    // Generate and save image
    try {
      const { ImageRenderer } = await import('../src/image-renderer');
      const renderer = new ImageRenderer();
      const imgData = renderer.renderToPng(compressed);
      
      const imgFilename = `rnn_paper_${label}_compressed.png`;
      writeFileSync(imgFilename, imgData);
      
      console.log(`  🖼️  Image saved: ${imgFilename}`);
      console.log(`     Size: ${Math.floor(imgData.length / 1024)} KB (${imgData.length} bytes)`);
      console.log('     Dimensions: 1024x1024 PNG');
      
      // Verify PNG signature
      if (imgData.length >= 8 && 
          imgData[0] === 137 && imgData[1] === 80 && imgData[2] === 78 && imgData[3] === 71 &&
          imgData[4] === 13 && imgData[5] === 10 && imgData[6] === 26 && imgData[7] === 10) {
        console.log('     ✓ Valid PNG signature');
      }

      // Calculate compression efficiency
      const charsPerKb = compressed.length / (imgData.length / 1024);
      console.log(`     Density: ${charsPerKb.toFixed(1)} chars/KB`);
    } catch (error) {
      console.error(`  ❌ Image rendering failed: ${error instanceof Error ? error.message : String(error)}`);
    }

    console.log();
  }

  console.log(sep);
  console.log('✅ Paper conversion complete!');
  console.log(sep);
  console.log();
  console.log('📊 Summary:');
  console.log(`  Input: ${paperPath}`);
  console.log("  Paper: 'On the difficulty of training Recurrent Neural Networks'");
  console.log('  Generated files:');
  console.log('    - rnn_paper_aggressive_compressed.txt + .png (30% compression)');
  console.log('    - rnn_paper_balanced_compressed.txt + .png (50% compression)');
  console.log('    - rnn_paper_light_compressed.txt + .png (70% compression)');
  console.log();
  console.log('💡 Use these PNG images with vision models for optical context compression!');
}

main().catch((error) => {
  console.error('Error:', error);
  process.exit(1);
});

File: rust/examples/paper_to_images.rs
use compression_prompt::{OutputFormat, StatisticalFilter, StatisticalFilterConfig};
use std::fs;
use std::path::Path;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("📄 Converting Paper to Compressed Images\n");

    // Read original paper
    let paper_path = "../benchmarks/datasets/arxiv_markdown/1505.07818.md";

    if !Path::new(paper_path).exists() {
        eprintln!("❌ Paper file not found: {}", paper_path);
        return Err("Paper file not found".into());
    }

    let text = fs::read_to_string(paper_path)?;

    println!("📝 Original paper:");
    println!("  File: {}", paper_path);
    println!("  Size: {} chars", text.len());
    println!("  Words: {}", text.split_whitespace().count());
    println!();

    let sep = "=".repeat(80);

    // Configuration for different compression levels
    let configs = vec![
        (0.3, "aggressive", "30% compression"),
        (0.5, "balanced", "50% compression (default)"),
        (0.7, "light", "70% compression"),
    ];

    for (ratio, label, description) in configs {
        println!("{}", sep);
        println!("Processing: {} - {}", label.to_uppercase(), description);
        println!("{}", sep);

        let config = StatisticalFilterConfig {
            compression_ratio: ratio,
            ..Default::default()
        };

        let filter = StatisticalFilter::new(config);
        let result = filter.compress_with_format(&text, OutputFormat::Image)?;

        println!("✅ Compression complete:");
        println!("  Original tokens: {}", result.original_tokens);
        println!("  Compressed tokens: {}", result.compressed_tokens);
        println!(
            "  Compression ratio: {:.1}%",
            result.compression_ratio * 100.0
        );
        println!("  Tokens saved: {}", result.tokens_removed);
        println!(
            "  Token savings: {:.1}%",
            (1.0 - result.compression_ratio) * 100.0
        );

        // Save compressed text
        let text_filename = format!("rnn_paper_{}_compressed.txt", label);
        fs::write(&text_filename, &result.compressed)?;
        println!("  📝 Text saved: {}", text_filename);

        // Save image
        if let Some(img_data) = &result.image_data {
            let img_filename = format!("rnn_paper_{}_compressed.png", label);
            fs::write(&img_filename, img_data)?;

            println!("  🖼️  Image saved: {}", img_filename);
            println!(
                "     Size: {} KB ({} bytes)",
                img_data.len() / 1024,
                img_data.len()
            );
            println!("     Dimensions: 1024x1024 PNG");

            // Verify PNG signature
            if img_data.len() >= 8 && img_data[0..8] == [137, 80, 78, 71, 13, 10, 26, 10] {
                println!("     ✓ Valid PNG signature");
            }

            // Calculate compression efficiency
            let chars_per_kb = result.compressed.len() as f32 / (img_data.len() as f32 / 1024.0);
            println!("     Density: {:.1} chars/KB", chars_per_kb);
        }

        println!();
    }

    println!("{}", sep);
    println!("✅ Paper conversion complete!");
    println!("{}", sep);
    println!();
    println!("📊 Summary:");
    println!("  Input: {}", paper_path);
    println!("  Paper: 'On the difficulty of training Recurrent Neural Networks'");
    println!("  Generated files:");
    println!("    - rnn_paper_aggressive_compressed.txt + .png (30% compression)");
    println!("    - rnn_paper_balanced_compressed.txt + .png (50% compression)");
    println!("    - rnn_paper_light_compressed.txt + .png (70% compression)");
    println!();
    println!("💡 Use these PNG images with vision models for optical context compression!");

    Ok(())
}
